framework/src/meshgenerators/MeshGenerator.C

```diff
@@ -20,15 +20,58 @@ MeshGenerator::validParams()
                         false,
                         "Whether or not to show mesh info after generating the mesh "
                         "(bounding box, element types, sidesets, nodesets, subdomains, etc)");
+  params.addParam<std::vector<std::string>>(
+      "selected_mesh_metadata_to_retain",
+      std::vector<std::string>(),
+      "Names of input mesh's metadata that are selected to retain.");
+  params.addParam<bool>("retain_all_input_mesh_metadata",
+                        false,
+                        "Whether to retain all the mesh metadata of the input mesh.");
 
   params.registerBase("MeshGenerator");
 
   return params;
 }
 
 MeshGenerator::MeshGenerator(const InputParameters & parameters)
-  : MooseObject(parameters), MeshMetaDataInterface(this), _mesh(_app.actionWarehouse().mesh())
+  : MooseObject(parameters),
+    MeshMetaDataInterface(this),
+    _mesh(_app.actionWarehouse().mesh()),
+    _selected_mesh_metadata_to_retain(
+        getParam<std::vector<std::string>>("selected_mesh_metadata_to_retain")),
+    _retain_all_input_mesh_metadata(getParam<bool>("retain_all_input_mesh_metadata"))
 {
+  if (isParamValid("input"))
+  {
+    const MeshGeneratorName input_name(getParam<MeshGeneratorName>("input"));
+    if (_retain_all_input_mesh_metadata)
+    {
+      if (!_selected_mesh_metadata_to_retain.empty())
+        paramError(
+            "selected_mesh_metadata_to_retain",
+            "This parameter should not be provided if retain_all_input_mesh_metadata is set true.");
+      const auto mesh_metadata_names = findMeshMetaData(input_name);
```

Comment:
so this constructor is run before the derived class'
so do we have the metadata set at this time?

Reply:
You are right. This would only work for metadata declared in initializer list. This is the case for many reactor module MGs. But I need to think about a more generalized approach.

Reply:
Moved to `MeshGenerator::generateInternal()`